File: curriculum/units/M12/code/python/query_loop.py
from __future__ import annotations

import asyncio
import copy
import inspect
from dataclasses import dataclass
from typing import Any, AsyncIterator, Awaitable, Callable, Protocol


Message = dict[str, Any]
QueryEvent = dict[str, Any]
ModelScript = Callable[
    [list[Message], "CancellationToken", int],
    Message | Awaitable[Message],
]


@dataclass(frozen=True)
class LoopTerminal:
    reason: str
    turns: int


class CancelledError(Exception):
    pass


class CancellationToken:
    def __init__(self) -> None:
        self._event = asyncio.Event()
        self.reason: str | None = None

    @property
    def cancelled(self) -> bool:
        return self._event.is_set()

    def cancel(self, reason: str) -> None:
        self.reason = reason
        self._event.set()

    def raise_if_cancelled(self) -> None:
        if self.cancelled:
            raise CancelledError(self.reason or "cancelled")

    async def wait(self) -> None:
        await self._event.wait()
        raise CancelledError(self.reason or "cancelled")


class TraceRecorder:
    def __init__(self) -> None:
        self.entries: list[dict[str, Any]] = []

    def record(self, event_type: str, **detail: Any) -> None:
        entry: dict[str, Any] = {"type": event_type}
        if detail:
            entry["detail"] = detail
        self.entries.append(entry)


class ModelPort(Protocol):
    def stream(
        self,
        messages: list[Message],
        token: CancellationToken,
    ) -> AsyncIterator[Message]:
        ...


class ToolPort(Protocol):
    name: str

    async def execute(
        self,
        input_data: dict[str, Any],
        token: CancellationToken,
    ) -> Any:
        ...


class ScriptedModel:
    def __init__(self, scripts: list[ModelScript]) -> None:
        self._scripts = scripts
        self.requests: list[list[Message]] = []

    async def stream(
        self,
        messages: list[Message],
        token: CancellationToken,
    ) -> AsyncIterator[Message]:
        token.raise_if_cancelled()
        request = copy.deepcopy(messages)
        self.requests.append(request)
        index = len(self.requests) - 1
        if index >= len(self._scripts):
            raise RuntimeError("unexpected model request")
        produced = self._scripts[index](request, token, index)
        message = await produced if inspect.isawaitable(produced) else produced
        yield copy.deepcopy(message)


class DurableConversation:
    def __init__(self, initial_messages: list[Message]) -> None:
        self._messages = copy.deepcopy(initial_messages)

    def consume(self, event: QueryEvent) -> None:
        if event["type"] in {"assistant", "tool_result"}:
            self._messages.append(copy.deepcopy(event["message"]))

    def snapshot(self) -> list[Message]:
        return copy.deepcopy(self._messages)


class QueryRun:
    """Owns Python's terminal side channel because async generators cannot return values."""

    def __init__(
        self,
        initial_messages: list[Message],
        model: ModelPort,
        tools: list[ToolPort],
        token: CancellationToken,
        trace: TraceRecorder,
        max_turns: int = 8,
    ) -> None:
        self._initial_messages = copy.deepcopy(initial_messages)
        self._model = model
        self._tools = {tool.name: tool for tool in tools}
        self._token = token
        self._trace = trace
        self._max_turns = max_turns
        self.terminal: LoopTerminal | None = None

    async def events(self) -> AsyncIterator[QueryEvent]:
        state_messages = copy.deepcopy(self._initial_messages)
        turns = 0
        try:
            while True:
                if self._token.cancelled:
                    yield {"type": "interruption", "phase": "before_request"}
                    self.terminal = LoopTerminal("aborted", turns)
                    return
                if turns >= self._max_turns:
                    self.terminal = LoopTerminal("max_turns", turns)
                    return

                turn = turns + 1
                request_messages = copy.deepcopy(state_messages)
                self._trace.record("producer.request.before_yield", turn=turn)
                yield {
                    "type": "request",
                    "turn": turn,
                    "messages": request_messages,
                }
                self._trace.record("producer.request.after_yield", turn=turn)

                assistant_messages: list[Message] = []
                try:
                    async for assistant in self._model.stream(
                        request_messages,
                        self._token,
                    ):
                        self._trace.record(
                            "producer.assistant.before_yield",
                            turn=turn,
                        )
                        yield {
                            "type": "assistant",
                            "message": copy.deepcopy(assistant),
                        }
                        self._trace.record(
                            "producer.assistant.after_yield",
                            turn=turn,
                        )
                        assistant_messages.append(copy.deepcopy(assistant))
                except CancelledError:
                    yield {"type": "interruption", "phase": "model"}
                    self.terminal = LoopTerminal("aborted", turns)
                    return
                except Exception as error:
                    yield {"type": "model_error", "message": str(error)}
                    self.terminal = LoopTerminal("model_error", turns)
                    return

                if self._token.cancelled:
                    yield {"type": "interruption", "phase": "model"}
                    self.terminal = LoopTerminal("aborted", turns)
                    return

                tool_uses = [
                    block
                    for message in assistant_messages
                    for block in message["blocks"]
                    if block["type"] == "tool_use"
                ]
                if not tool_uses:
                    self.terminal = LoopTerminal("completed", turn)
                    return

                tool_results: list[Message] = []
                for call in tool_uses:
                    if self._token.cancelled:
                        yield {"type": "interruption", "phase": "tool"}
                        self.terminal = LoopTerminal("aborted", turns)
                        return

                    tool = self._tools.get(call["name"])
                    is_error = False
                    try:
                        if tool is None:
                            raise RuntimeError(f"unknown tool: {call['name']}")
                        content = await tool.execute(call["input"], self._token)
                        self._token.raise_if_cancelled()
                    except Exception as error:
                        is_error = True
                        content = str(error)

                    result: Message = {
                        "role": "user",
                        "blocks": [
                            {
                                "type": "tool_result",
                                "toolUseId": call["id"],
                                "content": content,
                                "isError": is_error,
                            }
                        ],
                    }
                    self._trace.record(
                        "producer.tool_result.before_yield",
                        tool_use_id=call["id"],
                    )
                    yield {
                        "type": "tool_result",
                        "message": copy.deepcopy(result),
                    }
                    self._trace.record(
                        "producer.tool_result.after_yield",
                        tool_use_id=call["id"],
                    )
                    tool_results.append(result)

                    if self._token.cancelled:
                        yield {"type": "interruption", "phase": "tool"}
                        self.terminal = LoopTerminal("aborted", turn)
                        return

                state_messages = [
                    *state_messages,
                    *assistant_messages,
                    *tool_results,
                ]
                turns = turn
                self._trace.record(
                    "producer.state.continue",
                    turn=turn,
                    message_count=len(state_messages),
                )
        finally:
            self._trace.record("producer.loop.finally")


async def query_events(run: QueryRun, trace: TraceRecorder) -> AsyncIterator[QueryEvent]:
    inner = run.events()
    try:
        async for event in inner:
            yield event
        trace.record(
            "wrapper.normal_completion",
            reason=run.terminal.reason if run.terminal else None,
        )
    finally:
        await inner.aclose()
        trace.record("wrapper.finally")


def user_text(text: str) -> Message:
    return {"role": "user", "blocks": [{"type": "text", "text": text}]}


def assistant_text(text: str) -> Message:
    return {"role": "assistant", "blocks": [{"type": "text", "text": text}]}


def assistant_tool(
    tool_use_id: str,
    name: str,
    input_data: dict[str, Any],
) -> Message:
    return {
        "role": "assistant",
        "blocks": [
            {
                "type": "tool_use",
                "id": tool_use_id,
                "name": name,
                "input": input_data,
            }
        ],
    }


def find_tool_result(
    messages: list[Message],
    tool_use_id: str,
) -> dict[str, Any] | None:
    for message in messages:
        if message["role"] != "user":
            continue
        for block in message["blocks"]:
            if (
                block["type"] == "tool_result"
                and block["toolUseId"] == tool_use_id
            ):
                return block
    return None

File: curriculum/units/M12/code/typescript/query-loop.ts
export type TextBlock = {
  type: "text";
  text: string;
};

export type ToolUseBlock = {
  type: "tool_use";
  id: string;
  name: string;
  input: Record<string, unknown>;
};

export type ToolResultBlock = {
  type: "tool_result";
  toolUseId: string;
  content: unknown;
  isError: boolean;
};

export type UserMessage = {
  role: "user";
  blocks: Array<TextBlock | ToolResultBlock>;
};

export type AssistantMessage = {
  role: "assistant";
  blocks: Array<TextBlock | ToolUseBlock>;
};

export type Message = UserMessage | AssistantMessage;

export type QueryEvent =
  | {
      type: "request";
      turn: number;
      messages: readonly Message[];
    }
  | {
      type: "assistant";
      message: AssistantMessage;
    }
  | {
      type: "tool_result";
      message: UserMessage;
    }
  | {
      type: "interruption";
      phase: "before_request" | "model" | "tool";
    }
  | {
      type: "model_error";
      message: string;
    };

export type LoopTerminal = {
  reason: "completed" | "aborted" | "model_error" | "max_turns";
  turns: number;
};

export type TraceEntry = {
  type: string;
  detail?: Record<string, unknown>;
};

export class TraceRecorder {
  readonly entries: TraceEntry[] = [];

  record(type: string, detail?: Record<string, unknown>): void {
    this.entries.push(detail ? { type, detail } : { type });
  }
}

export interface ModelPort {
  stream(
    messages: readonly Message[],
    signal: AbortSignal,
  ): AsyncGenerator<AssistantMessage, void>;
}

export interface ToolPort {
  readonly name: string;
  execute(
    input: Record<string, unknown>,
    signal: AbortSignal,
  ): Promise<unknown>;
}

export type QueryParams = {
  initialMessages: readonly Message[];
  model: ModelPort;
  tools: readonly ToolPort[];
  signal: AbortSignal;
  trace: TraceRecorder;
  maxTurns?: number;
};

type LoopState = {
  messages: Message[];
  turns: number;
};

export async function* queryLoop(
  params: QueryParams,
): AsyncGenerator<QueryEvent, LoopTerminal> {
  const tools = new Map(params.tools.map((tool) => [tool.name, tool]));
  const maxTurns = params.maxTurns ?? 8;
  let state: LoopState = {
    messages: structuredClone([...params.initialMessages]),
    turns: 0,
  };

  try {
    while (true) {
      if (params.signal.aborted) {
        yield { type: "interruption", phase: "before_request" };
        return { reason: "aborted", turns: state.turns };
      }
      if (state.turns >= maxTurns) {
        return { reason: "max_turns", turns: state.turns };
      }

      const turn = state.turns + 1;
      const requestMessages = structuredClone(state.messages);
      params.trace.record("producer.request.before_yield", { turn });
      yield { type: "request", turn, messages: requestMessages };
      params.trace.record("producer.request.after_yield", { turn });

      const assistantMessages: AssistantMessage[] = [];
      try {
        for await (const assistant of params.model.stream(
          requestMessages,
          params.signal,
        )) {
          params.trace.record("producer.assistant.before_yield", { turn });
          yield { type: "assistant", message: structuredClone(assistant) };
          params.trace.record("producer.assistant.after_yield", { turn });
          assistantMessages.push(structuredClone(assistant));
        }
      } catch (error) {
        if (params.signal.aborted || error instanceof CancelledError) {
          yield { type: "interruption", phase: "model" };
          return { reason: "aborted", turns: state.turns };
        }
        const message = error instanceof Error ? error.message : String(error);
        yield { type: "model_error", message };
        return { reason: "model_error", turns: state.turns };
      }

      if (params.signal.aborted) {
        yield { type: "interruption", phase: "model" };
        return { reason: "aborted", turns: state.turns };
      }

      const toolUses = assistantMessages.flatMap((message) =>
        message.blocks.filter(
          (block): block is ToolUseBlock => block.type === "tool_use",
        ),
      );
      if (toolUses.length === 0) {
        return { reason: "completed", turns: turn };
      }

      const toolResults: UserMessage[] = [];
      for (const call of toolUses) {
        if (params.signal.aborted) {
          yield { type: "interruption", phase: "tool" };
          return { reason: "aborted", turns: state.turns };
        }

        const tool = tools.get(call.name);
        let content: unknown;
        let isError = false;
        try {
          if (!tool) throw new Error(`unknown tool: ${call.name}`);
          content = await tool.execute(call.input, params.signal);
          if (params.signal.aborted) {
            throw new CancelledError("cancelled during tool execution");
          }
        } catch (error) {
          isError = true;
          content = error instanceof Error ? error.message : String(error);
        }

        const result: UserMessage = {
          role: "user",
          blocks: [
            {
              type: "tool_result",
              toolUseId: call.id,
              content,
              isError,
            },
          ],
        };
        params.trace.record("producer.tool_result.before_yield", {
          toolUseId: call.id,
        });
        yield { type: "tool_result", message: structuredClone(result) };
        params.trace.record("producer.tool_result.after_yield", {
          toolUseId: call.id,
        });
        toolResults.push(result);

        if (params.signal.aborted) {
          yield { type: "interruption", phase: "tool" };
          return { reason: "aborted", turns: turn };
        }
      }

      state = {
        messages: [
          ...state.messages,
          ...assistantMessages,
          ...toolResults,
        ],
        turns: turn,
      };
      params.trace.record("producer.state.continue", {
        turn,
        messageCount: state.messages.length,
      });
    }
  } finally {
    params.trace.record("producer.loop.finally");
  }
}

export async function* query(
  params: QueryParams,
): AsyncGenerator<QueryEvent, LoopTerminal> {
  try {
    const terminal = yield* queryLoop(params);
    params.trace.record("wrapper.normal_completion", {
      reason: terminal.reason,
    });
    return terminal;
  } finally {
    params.trace.record("wrapper.finally");
  }
}

export async function drainWithTerminal(
  iterator: AsyncGenerator<QueryEvent, LoopTerminal>,
): Promise<{ events: QueryEvent[]; terminal: LoopTerminal }> {
  const events: QueryEvent[] = [];
  while (true) {
    const step = await iterator.next();
    if (step.done) return { events, terminal: step.value };
    events.push(step.value);
  }
}

export type ModelScript = (
  messages: readonly Message[],
  signal: AbortSignal,
  callIndex: number,
) => AssistantMessage | Promise<AssistantMessage>;

export class ScriptedModel implements ModelPort {
  readonly requests: Message[][] = [];
  private readonly scripts: readonly ModelScript[];

  constructor(scripts: readonly ModelScript[]) {
    this.scripts = scripts;
  }

  async *stream(
    messages: readonly Message[],
    signal: AbortSignal,
  ): AsyncGenerator<AssistantMessage, void> {
    if (signal.aborted) throw new CancelledError("model cancelled");
    const request = structuredClone([...messages]);
    this.requests.push(request);
    const script = this.scripts[this.requests.length - 1];
    if (!script) throw new Error("unexpected model request");
    yield structuredClone(
      await script(request, signal, this.requests.length - 1),
    );
  }
}

export class DurableConversation {
  private readonly messages: Message[];

  constructor(initialMessages: readonly Message[]) {
    this.messages = structuredClone([...initialMessages]);
  }

  consume(event: QueryEvent): void {
    if (event.type === "assistant" || event.type === "tool_result") {
      this.messages.push(structuredClone(event.message));
    }
  }

  snapshot(): readonly Message[] {
    return structuredClone(this.messages);
  }
}

export class CancelledError extends Error {}

export function userText(text: string): UserMessage {
  return { role: "user", blocks: [{ type: "text", text }] };
}

export function assistantText(text: string): AssistantMessage {
  return { role: "assistant", blocks: [{ type: "text", text }] };
}

export function assistantTool(
  id: string,
  name: string,
  input: Record<string, unknown>,
): AssistantMessage {
  return {
    role: "assistant",
    blocks: [{ type: "tool_use", id, name, input }],
  };
}

export function findToolResult(
  messages: readonly Message[],
  toolUseId: string,
): ToolResultBlock | undefined {
  for (const message of messages) {
    if (message.role !== "user") continue;
    for (const block of message.blocks) {
      if (block.type === "tool_result" && block.toolUseId === toolUseId) {
        return block;
      }
    }
  }
  return undefined;
}

export function waitUntilAborted(signal: AbortSignal): Promise<never> {
  if (signal.aborted) {
    return Promise.reject(new CancelledError("model cancelled"));
  }
  return new Promise((_, reject) => {
    signal.addEventListener(
      "abort",
      () => reject(new CancelledError("model cancelled")),
      { once: true },
    );
  });
}
